### PR TITLE
Bug 1734043: Added CSI provisioner to storage class

### PR DESCRIPTION
**Pull request by cloudbehl**
sc-form
Signed-off-by: cloudbehl <cloudbehl@gmail.com>

### CODE REVIEW

File: frontend/public/components/storage-class-form.tsx [hunk 1]
```diff
@@ -55,7 +57,105 @@ export class StorageClassForm_ extends React.Component<
...
+          name: 'Filesystem Type',
+          hintText: 'Ceph RBD filesystem type. Default set to ext4.',
+          required: true,
+        },
+      },
+    },
+    'cephfs.csi.ceph.com': {
+      title: 'Ceph FS',
+      provisioner: 'cephfs.csi.ceph.com',
+      documentationLink: 'https://rook.github.io/docs/rook/master/ceph-filesystem.html',
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: it's pointing to rook master, I think it's better to point to released version https://rook.github.io/docs/rook/v1.1/

File: frontend/public/components/storage-class-form.tsx [hunk 2]
```diff
@@ -55,7 +57,105 @@ export class StorageClassForm_ extends React.Component<
...
+          required: true,
+        },
+        'csi.storage.k8s.io/provisioner-secret-name': {
+          name: 'Provisioner Secret Name',
+          hintText: 'Provisioner Secret Name is the name of provisioner secret.',
+          required: true,
+        },
+        'csi.storage.k8s.io/provisioner-secret-namespace': {
+          name: 'Provisioner Secret Namespace',
+          hintText: 'Provisioner Secret Namespace is the namespace where Ceph is deployed.',
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: Provisioner Secret Namespace is the namespace where provisioner secret is created, this need to updated for other hintText also

File: frontend/public/components/storage-class-form.tsx [hunk 3]
```diff
@@ -55,7 +57,105 @@ export class StorageClassForm_ extends React.Component<
...
+          required: true,
+        },
+        imageFormat: {
+          name: 'Image Format',
+          hintText: 'Ceph RBD image format.',
+          required: true,
+        },
+        imageFeatures: {
+          name: 'Image Features',
+          hintText: 'Ceph RBD image features.',
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: currently, only layering is supported for image feature https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/rbd/storageclass.yaml#L28

File: frontend/public/components/storage-class-form.tsx [hunk 4]
```diff
@@ -55,7 +57,105 @@ export class StorageClassForm_ extends React.Component<
...
+        },
+        pool: {
+          name: 'Pool',
+          hintText: 'Ceph pool into which the RBD image shall be created',
+          required: true,
+        },
+        imageFormat: {
+          name: 'Image Format',
+          hintText: 'Ceph RBD image format.',
+          required: true,
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: only image format 2 is supported for RBD https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/rbd/storageclass.yaml#L25

File: frontend/public/components/storage-class-form.tsx [hunk 5]
```diff
@@ -55,7 +57,105 @@ export class StorageClassForm_ extends React.Component<
...
+        clusterID: {
+          name: 'Cluster ID',
+          hintText: 'Cluster ID is the namespace where Ceph is deployed.',
+          required: true,
+        },
+        fsName: {
+          name: 'Filesystem Name',
+          hintText: 'CephFS filesystem name into which the volume shall be created',
+          required: true,
+        },
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: cephfs supports optional pool parameter https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/cephfs/storageclass.yaml#L15

File: frontend/public/components/storage-class-form.tsx [hunk 6]
```diff
@@ -55,7 +57,113 @@ export class StorageClassForm_ extends React.Component<
...
+
+  CSIStorageTypes = {
+    'rbd.csi.ceph.com': {
+      title: 'Ceph RBD',
+      provisioner: 'rbd.csi.ceph.com',
+      documentationLink: 'https://rook.github.io/docs/rook/master/ceph-block.html',
+      parameters: {
+        clusterID: {
+          name: 'Cluster ID',
+          hintText: 'Cluster ID is the namespace where Ceph is deployed.',
```
Comment (pcbailey) [inline frontend/public/components/storage-class-form.tsx]: Just a couple of nitpick notes for all of the hint text properties to maintain consistency with the hint text from the other provisioners:
- Drop the period at the end
- Instead of using the format "`<Parameter name>` is` <description>`", just use the description. 

For example: 
'Cluster ID is the namespace where Ceph is deployed.'  --> 'The namespace where Ceph is deployed'

File: frontend/public/components/storage-class-form.tsx [hunk 7]
```diff
@@ -754,6 +828,7 @@ export class StorageClassForm_ extends React.Component<
     const dynamicContent = _.map(parameters, (parameter, key) => {
       const validationMsg = _.get(parameter, 'validationMsg', null);
       const isCheckbox = parameter.type === 'checkbox';
+      const selectedKey = `newStorageClass.parameters['${key}'].value`;
```
Comment (pcbailey) [inline frontend/public/components/storage-class-form.tsx]: Thanks for caring about the small things. This looks much cleaner. =)

File: frontend/public/components/storage-class-form.tsx [hunk 8]
```diff
@@ -1022,7 +1100,10 @@ export const ConnectedStorageClassForm = connect(
 )(StorageClassForm_);
 
 export const StorageClassForm = (props) => {
-  const resources = [{ kind: 'StorageClass', isList: true }];
+  const resources = [
+    { kind: 'StorageClass', isList: true, prop: 'sc' },
```
Comment (pcbailey) [inline frontend/public/components/storage-class-form.tsx]: Feel free to change this to use `referenceForModel`, as well.

File: frontend/public/components/storage-class-form.tsx [hunk 9]
```diff
@@ -55,7 +57,111 @@ export class StorageClassForm_ extends React.Component<
...
+      documentationLink: 'https://rook.github.io/docs/rook/master/ceph-filesystem.html',
+      parameters: {
+        clusterID: {
+          name: 'Cluster ID',
+          hintText: 'The namespace where Ceph is deployed',
+          required: true,
+        },
+        pool: {
+          name: 'Pool',
+          hintText: 'Ceph pool into which the RBD image shall be created',
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: not RBD image its `Ceph pool into which volume data shall be stored`

File: frontend/public/components/storage-class-form.tsx [hunk 10]
```diff
@@ -55,7 +57,111 @@ export class StorageClassForm_ extends React.Component<
...
+          required: true,
+        },
+        'csi.storage.k8s.io/node-stage-secret-name': {
+          name: 'Node Stage Secret Name',
+          hintText: 'The name of Node Stage secret',
+          required: true,
+        },
+        'csi.storage.k8s.io/node-stage-secret-namespace': {
+          name: 'Node Stage Secret Namespace',
+          hintText: 'The namespace where Ceph is deployed',
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: this need to be updated as https://github.com/openshift/console/pull/3172/files#diff-2d573e2f986c3bea9d6365a7a40ee904R147

File: frontend/public/components/storage-class-form.tsx [hunk 11]
```diff
@@ -55,7 +57,111 @@ export class StorageClassForm_ extends React.Component<
...
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: this need to be updated as https://github.com/openshift/console/pull/3172/files#diff-2d573e2f986c3bea9d6365a7a40ee904R147

File: frontend/public/components/storage-class-form.tsx [hunk 12]
```diff
@@ -55,7 +57,111 @@ export class StorageClassForm_ extends React.Component<
...
+    title: '',
+    provisioner: '',
+    parameters: {},
+  };
+
+  CSIStorageTypes = {
+    'rbd.csi.ceph.com': {
+      title: 'Ceph RBD',
+      provisioner: 'rbd.csi.ceph.com',
+      documentationLink: 'https://rook.github.io/docs/rook/master/ceph-block.html',
```
Comment (Madhu-1) [inline frontend/public/components/storage-class-form.tsx]: don't we need to fix this?

Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: @madhu We don't want to change the link every time in every release of UI. So is there a stable link I can provide?

File: frontend/public/components/storage-class-form.tsx (hunk 11 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: ack

File: frontend/public/components/storage-class-form.tsx (hunk 9 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: ack

**Comment by cloudbehl** [comment]
@Madhu-1 updated the PR. PTAL

**Comment by cloudbehl** [comment]
@spadgett PTAL

**Comment by spadgett** [comment]
The last test failure should be fixed by #3344

/retest

File: frontend/public/components/storage-class-form.tsx [hunk 13]
```diff
@@ -501,25 +551,48 @@ export class StorageClassForm_ extends React.Component<
...
 
+  csiProvisionerMap = (csiData) => {
+    const csiListedProvisioner: string[] = _.keys(this.CSIStorageTypes);
+    csiData.map((csi) => {
+      csiListedProvisioner.some((provisioner) => {
+        const hasProvisioner = getName(csi).includes(provisioner);
+        if (hasProvisioner) {
+          const provisionerData = _.cloneDeep(this.CSIStorageTypes[provisioner]);
+          provisionerData.provisioner = getName(csi);
+          this.storageTypes[getName(csi)] = provisionerData;
```
Comment (spadgett) [inline frontend/public/components/storage-class-form.tsx]: It's unexpected that calling `some` would have a side effect.

File: frontend/public/components/storage-class-form.tsx [hunk 14]
```diff
@@ -501,25 +551,48 @@ export class StorageClassForm_ extends React.Component<
...
     },
-  });
+  };
 
   reclaimPolicies = {
     Retain: 'Retain',
     Delete: 'Delete',
   };
 
+  csiProvisionerMap = (csiData) => {
```
Comment (spadgett) [inline frontend/public/components/storage-class-form.tsx]: Can you explain what this code is doing? It's a little hard to follow. This deserves a comment at least.

File: frontend/public/components/storage-class-form.tsx (hunk 13 above)
Comment (spadgett) [reply frontend/public/components/storage-class-form.tsx]: I'd think you would want to have `storageTypes` in state since the list is no longer status and it's something that is used in render.

File: frontend/public/components/storage-class-form.tsx [hunk 15]
```diff
@@ -55,7 +57,111 @@ export class StorageClassForm_ extends React.Component<
     this.previousName = '';
   }
 
-  storageTypes = Object.freeze({
```
Comment (spadgett) [inline frontend/public/components/storage-class-form.tsx]: Leave the static list of storage types as a frozen object constant, then add the dynamic ones into state.

File: frontend/public/components/storage-class-form.tsx [hunk 16]
```diff
@@ -55,7 +57,111 @@ export class StorageClassForm_ extends React.Component<
...
+  defaultProvisionerObj = {
+    title: '',
+    provisioner: '',
+    parameters: {},
+  };
+
+  CSIStorageTypes = {
```
Comment (spadgett) [inline frontend/public/components/storage-class-form.tsx]:
```suggestion
  CSIStorageTypes = Object.freeze({
```

File: frontend/public/components/storage-class-form.tsx (hunk 12 above)
Comment (Madhu-1) [reply frontend/public/components/storage-class-form.tsx]: @travisn can we use https://rook.io/docs/rook/v1.1/ link here?

Comment (travisn) [reply frontend/public/components/storage-class-form.tsx]: Rook doesn't currently have stable links except master, but you don't want to link to master docs since that will change. There is also a big warning at the top that you're looking at unreleased documentation. So you would need to use the v1.1 links.

File: frontend/public/components/storage-class-form.tsx (hunk 1 above)
Comment (travisn) [reply frontend/public/components/storage-class-form.tsx]: instead of rook.github.io, it's recommended to use https://rook.io/docs/rook/v1.1/ceph-filesystem.html

File: frontend/public/components/storage-class-form.tsx (hunk 12 above)
Comment (travisn) [reply frontend/public/components/storage-class-form.tsx]: Actually, does it make sense to link to the upstream documentation? This is for creating an OCS storage class, right? Shouldn't we be linking to downstream docs?

File: frontend/public/components/storage-class-form.tsx [hunk 17]
```diff
@@ -983,6 +1056,9 @@ export type StorageClassFormProps = {
   watchK8sList: (id: string, query: object, kind: object) => void;
   stopK8sWatch: (id: string) => void;
   k8s: any;
+  resources?: {
+    [key: string]: FirehoseResult | FirehoseResult<K8sResourceKind>;
```
[the remark this answers is not in the document]
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: Ack

File: frontend/public/components/storage-class-form.tsx (hunk 13 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: I declared storageTypes as another instance and freezed the other parameter objects.

File: frontend/public/components/storage-class-form.tsx (hunk 14 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: Added comments and used `every` instead of `some`. I want to break if the provisioner is found and forEach and map don't provide that. So that's why using every or some.

File: frontend/public/components/storage-class-form.tsx (hunk 15 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: Ack

File: frontend/public/components/storage-class-form.tsx (hunk 16 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: Ack

**Comment by cloudbehl** [comment]
@spadgett updated the PR.

File: frontend/public/components/storage-class-form.tsx (hunk 14 above)
Comment (spadgett) [reply frontend/public/components/storage-class-form.tsx]: `_.each` allows you to break by returning false. That would be more clear.

https://lodash.com/docs/4.17.15#forEach

**Comment by spadgett** [comment]
This appears to be a legitimate failure. It's a good practice to look at test errors before retesting.

```
FAIL __tests__/components/storage-class-form.spec.tsx
  ● StorageClassForm › renders type-specific settings when storage type is set

    TypeError: Cannot read property 'parameters' of undefined
```

**Comment by cloudbehl** [comment]
@spadgett I actually had a look when the test failed and tried to checking why the value is being undefined and but never hit a scenario that is being mentioned. I was trying to be optimistic and see if the same thing is happening again by running `/test frontend`. 

Not an expert on the test framework and but will try to fix it up.

File: frontend/public/components/storage-class-form.tsx (hunk 12 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: @travisn Will the v1.1 link be the same or do we need to change it every time in a release?

IMO, This is an upstream master code so it should not link downstream somewhere.

@spadgett opinion??

**Comment by cloudbehl** [comment]
@spadgett please have a look.

File: frontend/public/components/storage-class-form.tsx [hunk 18]
```diff
@@ -6,18 +6,20 @@ import * as fuzzy from 'fuzzysearch';
...
+import { getName } from '@console/shared';
 import {
   AsyncComponent,
   ButtonBar,
   Dropdown,
   NameValueEditorPair,
   history,
   ExternalLink,
+  Firehose,
+  FirehoseResult,
```
Comment (gnehapk) [inline frontend/public/components/storage-class-form.tsx]: Alpha imports

Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: ack

File: frontend/public/components/storage-class-form.tsx [hunk 19]
```diff
@@ -6,18 +6,20 @@ import * as fuzzy from 'fuzzysearch';
...
   NameValueEditorPair,
   history,
-  ExternalLink,
 } from './utils';
-import { Firehose } from './utils/firehose';
-import { k8sCreate } from './../module/k8s';
+import { k8sCreate, referenceForModel, K8sResourceKind } from './../module/k8s';
 import * as k8sActions from '../actions/k8s';
-import { StorageClassModel } from './../models';
+import { StorageClassModel, CSIDriverModel } from './../models';
```
Comment (gnehapk) [inline frontend/public/components/storage-class-form.tsx]: Alpha imports

File: frontend/public/components/storage-class-form.tsx [hunk 20]
```diff
@@ -6,18 +6,20 @@ import * as fuzzy from 'fuzzysearch';
...
+  ExternalLink,
+  Firehose,
+  FirehoseResult,
   NameValueEditorPair,
   history,
-  ExternalLink,
 } from './utils';
-import { Firehose } from './utils/firehose';
-import { k8sCreate } from './../module/k8s';
+import { k8sCreate, referenceForModel, K8sResourceKind } from './../module/k8s';
```
Comment (gnehapk) [inline frontend/public/components/storage-class-form.tsx]: Alpha imports

**Comment by cloudbehl** [comment]
@gnehapk PTAL

**Comment by cloudbehl** [comment]
@spadgett please review it once again.

File: frontend/public/components/storage-class-form.tsx [hunk 21]
```diff
@@ -508,23 +560,54 @@ export class StorageClassForm_ extends React.Component<
     Delete: 'Delete',
   };
 
+  // Accepts a list of CSI provisioners and it checks if the
+  // provisioner is listed in CSIStorageTypes object
+  // if yes then return the provisioner with parameters that
+  // needs to be filled by user.
+  csiProvisionerMap = (csiData) => {
```
Comment (rawagner) [inline frontend/public/components/storage-class-form.tsx]: Can't this be simplified to following ?

```
csiProvisionerMap = (csiData) => {
  csiData.map((csi) => {
    const provisionerData = _.cloneDeep(this.CSIStorageTypes[getName(csi)] || this.defaultProvisionerObj);
    provisionerData.title = getName(csi);
    provisionerData.provisioner = getName(csi);
    this.storageTypes[getName(csi)] = provisionerData;
});
```

File: frontend/public/components/storage-class-form.tsx (hunk 12 above)
Comment (rawagner) [reply frontend/public/components/storage-class-form.tsx]: I'd also say that 1.1 https://rook.io/docs/rook/v1.1/ is better than master. Ideally the link should come from some config map populated by rook operator

File: frontend/public/components/storage-class-form.tsx (hunk 21 above)
Comment (cloudbehl) [reply frontend/public/components/storage-class-form.tsx]: Nope, it will not work. As the provisioner name is prefixed with namespace name but for listedCSI I am using a non prefixed name like this `rbd.csi.ceph.com`.
